zero/src/rpc/jumpdest.rs

```diff
@@ -0,0 +1,512 @@
+                // We reverse the stack, so the order matches our assembly code.
+                let evm_stack: Vec<_> = entry.stack.as_ref().unwrap().iter().rev().collect();
+                let operands = 1;
+                if evm_stack.len() < operands {
+                    trace!( "Opcode {op} expected {operands} operands at the EVM stack, but only {} were found.", evm_stack.len() );
+                    continue;
+                }
+                let [jump_target, ..] = evm_stack[..] else {
+                    unreachable!()
+                };
+
+                prev_jump = Some(*jump_target);
+            }
+            "JUMPI" => {
+                prev_jump = None;
+                ensure!(entry.stack.as_ref().is_some(), "No evm stack found.");
+                // We reverse the stack, so the order matches our assembly code.
+                let evm_stack: Vec<_> = entry.stack.as_ref().unwrap().iter().rev().collect();
+                let operands = 2;
+                if evm_stack.len() < operands {
+                    trace!( "Opcode {op} expected {operands} operands at the EVM stack, but only {} were found.", evm_stack.len());
+                    continue;
+                };
+
+                let [jump_target, condition, ..] = evm_stack[..] else {
+                    unreachable!()
+                };
+                let jump_condition = condition.is_zero().not();
+
+                if jump_condition {
+                    prev_jump = Some(*jump_target)
+                }
+            }
+            "JUMPDEST" => {
+                let mut jumped_here = false;
+
+                if let Some(jmp_target) = prev_jump {
+                    jumped_here = jmp_target == U256::from(entry.pc);
+                }
+                prev_jump = None;
+
+                if jumped_here.not() {
+                    trace!(
+                        "{op}: JUMPDESTs at offset {} was reached through fall-through.",
+                        entry.pc
+                    );
+                    continue;
+                }
+
+                let jumpdest_offset = TryInto::<usize>::try_into(entry.pc);
+                if jumpdest_offset.is_err() {
+                    trace!(
+                        "{op}: Could not cast offset {} to usize {}.",
+                        entry.pc,
+                        usize::MAX
+                    );
+                    continue;
+                }
+                jumpdest_table.insert(*code_hash, *ctx, jumpdest_offset.unwrap());
+            }
+            "EXTCODECOPY" | "EXTCODESIZE" => {
+                prev_jump = None;
+                next_ctx_available += 1;
+            }
+            _ => {
+                prev_jump = None;
+            }
+        }
+    }
+    Ok((jumpdest_table, code_db))
+}
+
+/// This module exists as a workaround for parsing `StructLog`.  The `error`
+/// field is a string in Alloy but an object in Erigon.
+pub mod structlogprime {
```

Comment:
I'm sure there's a better way to do this.

Can you tell me more about what the `error` field emitted by `Erigon` looks like, and what we want to do with it?

Reply:
I submitted a PR upstream: https://github.com/erigontech/erigon/pull/12089

For my purposes nothing, hence my solution is to `skip` it. I just need a `&[StructLog]` to work with.

Reply:
https://github.com/0xPolygonZero/zk_evm/pull/681